keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,779 @@
+after the job finished.
+
+### Transition of "status.conditions"
+
+When the job with successPolicies is submitted, the job `status.conditions` transits in the following:
+Note that the Job doesn't have an actual `Running` condition in the `status.conditions`.
+
+```mermaid
+stateDiagram-v2
+    [*] --> Running
+    Running --> Failed: Exceeded backoffLimit\n or Matched FailurePolicy
+    Failed --> [*]
+    Running --> SuccessCriteriaMet: Matched SuccessPolicy
+    SuccessCriteriaMet --> Complete: All pods were terminated by successPolicies
+    Complete --> [*]
+```
+
+It means that the job's `.status.conditions` follows the following rules:  
+
+- The job could have both `SuccessCriteriaMet=true` and `Complete=true` conditions.
+- The job couldn't have both `Failed=true` and `SuccessCriteriaMet=true` conditions.
+- The job couldn't have both `Failed=true` and `Complete=true` conditions.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with other terminating policies such as the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+, backoffLimit, and [backoffLimitPerIndex](https://kubernetes.io/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
+in the following situations:
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+- when the job meets the successPolicy and the number of failed pods per indexes exceeds `backoffLimitPerIndex` in all indexes.
+
+To avoid the above conflicts, terminating policies are evaluated the first before successPolicies.
+This means that the terminating policies are respected rather than the successPolicies.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test cases:
+  - tests for Defaulting and Validating
+  - verify whether a job has a SuccessCriteriaMet condition if the job meets to successPolicy and some indexes fail.
+  - verify whether a job has both complete and SuccessCriteriaMet conditions 
+  if the job meets to successPolicy and all pods are terminated
+  - verify whether a job has a failed condition if the job meets to both successPolicy and terminating policies
+  in the same reconcile cycle
+
+- `k8s.io/kubernetes/pkg/controller/job`: `2 February 2024` - `91.5%`
+- `k8s.io/kubernetes/pkg/apis/batch/validation`: `2 February 2024` - `98.0%`
+
+##### Integration tests
+
+- Test scenarios:
+  - enabling, disabling and re-enabling of the `JobSuccessPolicy` feature gate
+  - handling of successPolicy when all indexes succeeded
+  - handling of the `.spec.successPolicy.criteria.succeededIndexes` when some indexes fail
```

Comment:
```suggestion
  - handling of the `.spec.successPolicy.criteria.succeededIndexes` when some indexes remain pending
```
The concept of failed index is only introduced by backoffLimitPerIndex, and for that you have a dedicated point below

Reply:
Yes, you're right.

Reply:
Yes, you're right.

Reply:
Done.